src/sources/util/tcp.rs

```diff
@@ -131,6 +134,7 @@ pub trait TcpSource: Clone + Send + 'static {
                     Ok(())
                 })
                 .inspect(|_| trigger.cancel());
+            // todo signal when shutdown is complete
```

Comment:
When I first read your [proposal](https://github.com/timberio/vector/issues/1091), @LucioFranco, I didn't understand why you suggested that the Sources would have "handle" that would notify the shutdown coordinator that their shutdown is complete by dropping the handle, so I went with the approach coded here of using a Trigger object directly, figuring the Sources could just signal the Trigger whenever their shutdown is complete.

Now, having waded into the code a bit deeper, I think I understand better.  The issue is that a single source can spawn multiple futures in parallel, so knowing once all child futures have been fully cleaned up is difficult.  If instead we had a reference-counted object passed into each spawned future, then as the various future chains resolve at shutdown we'd know all sub-tasks have completed when the refcount of the handle object goes to zero.  The handle object's destructor could then signal a Trigger object to notify the shutdown coordinator that this source's shutdown is complete.

So now I'm thinking I should replace the ShutdownSignals.shutdown_complete Trigger object with a Rc<TriggerHandle> that contains the equivalent of the current shutdown_complete Trigger object and overloads the Drop trait to call .cancel() on the Trigger.

Is this in line with what you were thinking?

Reply:
Yes, this is some what along the lines of what I was thinking. This should be very similar to a waitgroup in go. I am not 100% sure on the API yet but we can find ways to for example wrap spawn so when you spawn a future it will get a new bumped ref counted trigger with it. This way we can even track how many subtasks per source are left.

Happy to talk more through this via slack as well :)